src/liballoc/collections/vec_deque.rs

```diff
@@ -202,6 +202,22 @@ impl<T> VecDeque<T> {
                                  len);
     }
 
+    /// Copies all values from `src` to `self`, wrapping around if needed.
+    /// Assumes capacity is sufficient.
+    #[inline]
+    unsafe fn copy_slice(&mut self, src: &[T]) {
+        let dst_high_ptr = self.ptr().add(self.head);
+        let dst_high_len = self.cap() - self.head;
+
+        let split = cmp::min(src.len(), dst_high_len);
+        let (src_high, src_low) = src.split_at(split);
+
+        ptr::copy_nonoverlapping(src_high.as_ptr(), dst_high_ptr, src_high.len());
+        ptr::copy_nonoverlapping(src_low.as_ptr(), self.ptr(), src_low.len());
+
+        self.head = self.wrap_add(self.head, src.len());
```

Comment:
Can the `head` wrap over the tail ? Maybe it might be worth it to comment what the implementation is doing and why it works. Also maybe it might be worth it to add a `debug_assert!` that checks that the `head` does not wrap over the `tail`.

Reply:
Because the capacity is sufficient to hold `src.len()` new elements, this can't wrap over the tail.

Reply:
Duh, `head` points to the first element where data can be written, and `tail` to the first element that can be read. So this makes sense now, thanks.